lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -73,6 +78,522 @@ struct DiagRestore {
+      }
+    }
+
+    for (unsigned i = 0; i < NumStreams; ++i) {
+      OS << "Outputs affected by inputs as a table of bitmasks for stream " << i
+         << ":\n";
+      uint8_t InputVectors =
+          llvm::RoundUpToAlignment(NumInputSigScalars, 4) / 4;
+      uint8_t OutputVectors =
+          llvm::RoundUpToAlignment(NumOutputSigScalars[i], 4) / 4;
+      const PSVDependencyTable Table(InputToOutputTable[i].data(), InputVectors,
+                                     OutputVectors);
+      std::string OutputSetName = "Outputs";
+      OutputSetName += "[" + std::to_string(i) + "]";
+      Table.Print(OS, "Inputs", OutputSetName.c_str());
+    }
+
+    if (ShaderStage == PSVShaderKind::Hull ||
+        ShaderStage == PSVShaderKind::Mesh) {
+      OS << "Patch constant outputs affected by inputs as a table of "
+            "bitmasks:\n";
+      uint8_t InputVectors =
+          llvm::RoundUpToAlignment(NumInputSigScalars, 4) / 4;
+      uint8_t OutputVectors =
+          llvm::RoundUpToAlignment(NumPCOrPrimSigScalars, 4) / 4;
+      const PSVDependencyTable Table(InputToPCOutputTable.data(), InputVectors,
+                                     OutputVectors);
+      Table.Print(OS, "Inputs", "PatchConstantOutputs");
+    } else if (ShaderStage == PSVShaderKind::Domain) {
+      OS << "Outputs affected by patch constant inputs as a table of "
+            "bitmasks:\n";
+      uint8_t InputVectors =
+          llvm::RoundUpToAlignment(NumPCOrPrimSigScalars, 4) / 4;
+      uint8_t OutputVectors =
+          llvm::RoundUpToAlignment(NumOutputSigScalars[0], 4) / 4;
+      const PSVDependencyTable Table(PCInputToOutputTable.data(), InputVectors,
+                                     OutputVectors);
+      Table.Print(OS, "PatchConstantInputs", "Outputs");
+    }
+  }
+};
+SimpleViewIDState::SimpleViewIDState(std::vector<uint32_t> &Data,
+                                     PSVShaderKind SK, bool UsesViewID) {
+  if (Data.empty())
+    return;
+
+  unsigned Offset = 0;
+  // #Inputs.
+  NumInputSigScalars = Data[Offset++];
+  unsigned NumStreams =
+      SK == PSVShaderKind::Geometry ? DXIL::kNumOutputStreams : 1;
+  for (unsigned i = 0; i < NumStreams; i++) {
+    // #Outputs for stream StreamId.
+    unsigned OutputScalars = Data[Offset++];
+    NumOutputSigScalars[i] = OutputScalars;
+    unsigned MaskDwords = llvm::RoundUpToAlignment(OutputScalars, 32) / 32;
+    if (UsesViewID) {
+      ViewIDOutputMask[i] = {&(Data[Offset]), MaskDwords};
+      Offset += MaskDwords;
+    }
+    unsigned TabSize = MaskDwords * NumInputSigScalars;
+    if (TabSize == 0)
+      continue;
+    unsigned AlignedTabSize =
+        MaskDwords * llvm::RoundUpToAlignment(NumInputSigScalars, 4);
+    InputToOutputTable[i] = std::vector<uint32_t>(AlignedTabSize, 0);
+    memcpy(InputToOutputTable[i].data(), &(Data[Offset]), TabSize * 4);
+    Offset += TabSize;
+  }
+  if (SK == PSVShaderKind::Hull || SK == PSVShaderKind::Mesh) {
```

Comment:
These Hull/Mesh and Domain cases are the only places where `NumPCOrPrimSigScalars and the contents of `NumOutputSigSAcalars` are used without converting to vectors, but the values are retrieved in this function and could be temporarily saved for use here.